README.md
Platform drivers of Celestica products for the SONiC project

debian/changelog
sonic-cel-platform-modules (1.1) unstable; urgency=low

  * Add Phalanx platform.
  * Add Fishbone48 platform.
  * Add Fishbone32 platform.

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Mon, 03 Sep 2018 17:10:10 +0700

sonic-cel-platform-modules (1.0) unstable; urgency=low

  * Add Silverstone platform.

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Mon, 06 Aug 2018 12:18:15 +0700

sonic-cel-platform-modules (0.9) unstable; urgency=low

  * Add Questone2 platform.

 -- Wirut Getbamrung <wgetbumr@celestica.com>  Tue, 31 Jul 2018 10:40:00 +0700

sonic-cel-platform-modules (0.8) unstable; urgency=low

  * Add dx010 platform fan led control.
  * The platform gpio init moved to kernel space. 

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Mon, 1 Jul 2018 11:09:13 +0700

sonic-cel-platform-modules (0.7) unstable; urgency=low

  * Add dx010 plaform gpio sysfs exported when module load.

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Wed, 21 Jun 2018 13:29:05 +0700

sonic-cel-platform-modules (0.6) unstable; urgency=low

  * Remove unused port-mode switch script. This should be done by hwsku config script.
  * Add script to turn off QSFP low power mode when boot up.

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Wed, 26 July 2017 10:43:00 +0700

sonic-cel-platform-modules (0.5) unstable; urgency=low

  * Add port-mode switch script to support 100G 50G 10G_50G qsfp modes.
  * Fix garbage data when using sfputil to read QSFP-transceiver's eeprom.
  * Fix incorrect endian in eeprom read word data.

 -- Pradchaya Phucharoen <pphuchar@celestica.com>  Tue, 18 July 2017 11:30:00 +0700

sonic-cel-platform-modules (0.4) unstable; urgency=low

  * Add support for DX010's fancontrol, automatic run-up and FIX bug lpmod

 -- Pariwat Leamsumran <pleamsum@celestica.com>  Thu, 14 June 2017 16:25:00 +0700

sonic-cel-platform-modules (0.3) unstable; urgency=low

  * Add support for DX010's DPS800

 -- Abhisit Sangjan <asang@celestica.com>  Thu, 29 May 2017 19:23:00 +0700

sonic-cel-platform-modules (0.2) unstable; urgency=low

  * Add support for DX010's LM75B, Watchdog and EMC2305

 -- Abhisit Sangjan <asang@celestica.com>  Thu, 25 May 2017 15:26:00 +0700

sonic-cel-platform-modules (0.1) unstable; urgency=low

  * Initial release

 -- Abhisit Sangjan <asang@celestica.com>  Mon, 2 May 2017 14:47:00 +0700


dx010/modules/dx010-fan-led.c
/*
 * dx010-fan-led.c - Kernel module for Seastone DX010 fan LEDs
 *
 * Copyright (C) 2018 Celestica Corp.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <linux/module.h>
#include <linux/workqueue.h>
#include <linux/gpio.h>
#include <linux/kernel.h>
#include <linux/err.h>
#include <linux/printk.h>

#define MODULE_NAME "dx010-fan-led"

#define NUM_FAN 5
#define NUM_PSU 2
#define UPDATE_MS 100
#define CHECK_MS 1000
#define CHECK_COUNTER  CHECK_MS/UPDATE_MS


struct fan_gpio {
    unsigned char presence_pin;
    unsigned char red_led_pin;
    unsigned char yel_led_pin;
    int prs;
};

struct psu_gpio {
    unsigned char absence_pin;
    unsigned char powerok_pin;
};

static int count;
static struct workqueue_struct *dx010_fan_led_workqueue;
static struct delayed_work fan_update;
static struct gpio_chip *gc;
static struct fan_gpio fan_gpios[NUM_FAN] = {
  {.presence_pin = 10,
   .red_led_pin = 29,
   .yel_led_pin = 30,
   .prs = 1
  },
  {.presence_pin = 11,
   .red_led_pin = 31,
   .yel_led_pin = 32,
   .prs = 1
  },
  {.presence_pin = 12,
   .red_led_pin = 33,
   .yel_led_pin = 34,
   .prs = 1
  },
  {.presence_pin = 13,
   .red_led_pin = 35,
   .yel_led_pin = 36,
   .prs = 1
  },
  {.presence_pin = 14,
   .red_led_pin = 37,
   .yel_led_pin = 38,
   .prs = 1
  }
};

static struct psu_gpio psu_gpios[NUM_PSU] = {
  {.absence_pin = 27,
   .powerok_pin = 22
  },
  {.absence_pin = 28,
   .powerok_pin = 25,
  }
};

static int chip_match_name(struct gpio_chip *chip, void *data){
  return !strcmp(chip->label, data);
}

static void init_fan_gpio(void){
  int index;
  for( index=0; index < NUM_FAN; index++){
    gpio_request(gc->base+fan_gpios[index].presence_pin, NULL);
    gpio_request(gc->base+fan_gpios[index].red_led_pin, NULL);
    gpio_request(gc->base+fan_gpios[index].yel_led_pin, NULL);
    gpio_direction_input(gc->base+fan_gpios[index].presence_pin);
    gpio_direction_output(gc->base+fan_gpios[index].red_led_pin, 1);
    gpio_direction_output(gc->base+fan_gpios[index].yel_led_pin, 0);
  }
}

static void init_psu_gpio(void){
  int index;
  for( index=0; index < NUM_PSU; index++){
    gpio_request(gc->base+psu_gpios[index].absence_pin, NULL);
    gpio_request(gc->base+psu_gpios[index].powerok_pin, NULL);
    gpio_direction_input(gc->base+psu_gpios[index].absence_pin);
    gpio_direction_input(gc->base+psu_gpios[index].powerok_pin);
    gpio_export(gc->base+psu_gpios[index].absence_pin,0);
    gpio_export(gc->base+psu_gpios[index].powerok_pin,0);
  }
}

static void update_led(void){
  int index;
  int presence;
  int delay_ms;
  struct gpio_chip *find;

  /* Check for gpio chip exist every CHECK_MS */
  if(count <= 0){
    find = gpiochip_find("pca9505", chip_match_name);
    if(find){
      /* Skip check routine, Update the leds */
      if(gc != find){
        gc = find;
        init_fan_gpio();
        init_psu_gpio();
      }
      count = CHECK_COUNTER;
      delay_ms = UPDATE_MS;
    }else{
      /* Put the check routine to the queue */
      delay_ms = CHECK_MS;
    }
  }else{
    for (index=0; index < NUM_FAN; index++){
      presence = !gpio_get_value_cansleep(gc->base+fan_gpios[index].presence_pin);
      if( presence != fan_gpios[index].prs ){
        if( presence ){
          gpio_set_value_cansleep(gc->base+fan_gpios[index].red_led_pin, 1);
          gpio_set_value_cansleep(gc->base+fan_gpios[index].yel_led_pin, 0);
        }else{
          gpio_set_value_cansleep(gc->base+fan_gpios[index].red_led_pin, 0);
          gpio_set_value_cansleep(gc->base+fan_gpios[index].yel_led_pin, 1);
        }
      }
      fan_gpios[index].prs = presence;
    }
    count--;
  }
  queue_delayed_work(dx010_fan_led_workqueue, &fan_update,
                     msecs_to_jiffies(delay_ms));
}

static void deinit_fan_gpio(void){
  int index;
  for( index=0; index < NUM_PSU; index++){
    gpio_free(gc->base+fan_gpios[index].presence_pin);
    gpio_free(gc->base+fan_gpios[index].red_led_pin);
    gpio_free(gc->base+fan_gpios[index].yel_led_pin);
  }
}

static void deinit_psu_gpio(void){
  int index;
  for( index=0; index < NUM_PSU; index++){
    gpio_unexport(gc->base+psu_gpios[index].absence_pin);
    gpio_unexport(gc->base+psu_gpios[index].powerok_pin);
    gpio_free(gc->base+psu_gpios[index].absence_pin);
    gpio_free(gc->base+psu_gpios[index].powerok_pin);
  }
}

static int __init dx010_fan_led_init(void){
  count = 0;
  gc = NULL;
  dx010_fan_led_workqueue = create_singlethread_workqueue(MODULE_NAME);
  if (IS_ERR(dx010_fan_led_workqueue)) {
    printk(KERN_INFO "failed to inittialize workqueue\n");
    return PTR_ERR(dx010_fan_led_workqueue);
  }

  INIT_DELAYED_WORK(&fan_update, update_led);
  queue_delayed_work(dx010_fan_led_workqueue, &fan_update,
                     msecs_to_jiffies(UPDATE_MS));
  return 0;
}

static void __exit dx010_fan_led_exit(void){
  cancel_delayed_work_sync(&fan_update);
  destroy_workqueue(dx010_fan_led_workqueue);
  deinit_fan_gpio();
  deinit_psu_gpio();
  dx010_fan_led_workqueue = NULL;
  gc = NULL;
}

module_init(dx010_fan_led_init);
module_exit(dx010_fan_led_exit);


MODULE_LICENSE("GPL");
MODULE_AUTHOR("Celestica Inc.");
MODULE_DESCRIPTION("DX010 fan led control");